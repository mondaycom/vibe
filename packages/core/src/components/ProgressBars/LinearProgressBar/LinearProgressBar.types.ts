import { SIZES } from "../../../constants";

export type LinearProgressBarStyle = "primary" | "secondary" | "positive" | "negative" | "warning" | "none";

export type LinearProgressBarSize = (typeof SIZES)[keyof typeof SIZES];
