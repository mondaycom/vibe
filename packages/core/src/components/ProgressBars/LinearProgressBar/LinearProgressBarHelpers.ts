import { toNumber } from "es-toolkit/compat";
import { type SIZES } from "./../../../constants/sizes";
import cx from "classnames";
import styles from "./LinearProgressBar.module.scss";

export type Size = (typeof SIZES)[keyof typeof SIZES];

export const calculatePercentage = (value: number, min: number, max: number) => {
  const valuePercentage = (toNumber(value - min) / toNumber(max - min)) * 100;
  return valuePercentage > 100 ? 100 : valuePercentage;
};

export const getProgressBarClassNames = (value: number) => {
  return cx(styles.progressBar, { [styles.completed]: value >= 100 });
};
