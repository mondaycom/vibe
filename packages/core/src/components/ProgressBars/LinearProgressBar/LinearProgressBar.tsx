import React, { forwardRef, useMemo } from "react";
import cx from "classnames";
import { SIZES } from "../../../constants";
import { getStyle } from "../../../helpers/typesciptCssModulesHelper";
import PercentageLabel from "../PercentageLabel/PercentageLabel";
import {
  ProgressBarStyle as ProgressBarStyleEnum,
  ProgressBarType as ProgressBarTypeEnum
} from "./LinearProgressBarConstants";
import { LinearProgressBarSize, LinearProgressBarStyle } from "./LinearProgressBar.types";
import { calculatePercentage, getProgressBarClassNames } from "./LinearProgressBarHelpers";
import Bar from "./Bar/Bar";
import { VibeComponent, VibeComponentProps, withStaticProps } from "../../../types";
import { ComponentDefaultTestId } from "../../../tests/constants";
import { getTestId } from "../../../tests/test-ids-utils";
import styles from "./LinearProgressBar.module.scss";

export interface LinearProgressBarProps extends VibeComponentProps {
  /**
   * Determines the visual style of the progress bar.
   */
  barStyle?: LinearProgressBarStyle;
  /**
   * The minimum value of the progress bar.
   */
  min?: number;
  /**
   * The maximum value of the progress bar.
   */
  max?: number;
  /**
   * The current progress value.
   */
  value?: number;
  /**
   * The secondary progress value.
   */
  valueSecondary?: number;
  /**
   * If true, enables animation effects.
   */
  animated?: boolean;
  /**
   * The size of the progress bar.
   */
  size?: LinearProgressBarSize;
  /**
   * If true, displays the progress percentage.
   */
  indicateProgress?: boolean;
  /**
   * If true, enables multiple progress bars.
   * **Note:** `value`, `valueSecondary`, and `barStyle` will not be used.
   */
  multi?: boolean;
  /**
   * An array of bar values and colors for multi-bar mode.
   */
  multiValues?: {
    /**
     * The progress value for a bar.
     */
    value?: number;
    /**
     * The bar color in hex format (`#000000` - `#ffffff`).
     */
    color?: string;
  }[];
  /**
   * The ARIA label for the progress bar.
   */
  ariaLabel?: string;
  /**
   * If true, makes the progress bar span the full container width.
   */
  fullWidth?: boolean;
}

const LinearProgressBar: VibeComponent<LinearProgressBarProps, HTMLDivElement> & {
  styles?: typeof ProgressBarStyleEnum;
  barStyles?: typeof ProgressBarStyleEnum;
  types?: typeof ProgressBarTypeEnum;
  barTypes?: typeof ProgressBarTypeEnum;
  sizes?: typeof SIZES;
} = forwardRef(
  (
    {
      min = 0,
      max = 100,
      value = 0,
      valueSecondary = 0,
      animated = true,
      barStyle = "primary",
      className,
      size = "small",
      indicateProgress = false,
      multi = false,
      multiValues = [],
      ariaLabel = "",
      id,
      fullWidth = false,
      "data-testid": dataTestId
    }: LinearProgressBarProps,
    ref
  ) => {
    const wrapperClassName = useMemo(() => {
      return cx(
        styles.wrapper,
        {
          [getStyle(styles, size.toString())]: size,
          [styles.fullWidth]: fullWidth
        },
        className
      );
    }, [size, fullWidth, className]);

    const valuePercentage = useMemo(() => {
      if (multi) {
        const firstValue = multiValues && multiValues.length && multiValues[0].value;
        if (firstValue === null || firstValue === undefined) return 0;
        return calculatePercentage(firstValue, min, max);
      }
      if (value === null || value === undefined) return 0;
      return calculatePercentage(value, min, max);
    }, [value, min, max, multi, multiValues]);

    const renderMultiBars = useMemo(() => {
      if (!multi) return null;
      return (
        <>
          {[...multiValues].reverse().map(({ value: baseValue, color }, i) => (
            <Bar
              className={getProgressBarClassNames(baseValue)}
              barStyle="none"
              value={baseValue}
              animated={animated}
              type="primary"
              color={color}
              min={min}
              max={max}
              id={`bar_${color}_${i}`}
              key={`bar_${color}_${i}`}
            />
          ))}
        </>
      );
    }, [min, max, animated, multiValues, multi]);

    const renderPercentage = indicateProgress ? (
      <PercentageLabel forElement="linear-progress-bar" value={valuePercentage} />
    ) : null;

    const renderBaseBars = !multi ? (
      <>
        <Bar
          className={getProgressBarClassNames(value)}
          barLabelName={ariaLabel}
          barStyle={barStyle}
          value={valueSecondary}
          animated={animated}
          type="secondary"
          min={min}
          max={max}
          data-testid={ComponentDefaultTestId.BAR_SECONDARY}
        />
        <Bar
          className={getProgressBarClassNames(value)}
          barStyle={barStyle}
          value={value}
          animated={animated}
          type="primary"
          min={min}
          max={max}
          data-testid={ComponentDefaultTestId.BAR_PRIMARY}
        />
      </>
    ) : null;

    return (
      <div
        className={wrapperClassName}
        ref={ref}
        id={id}
        data-testsid={dataTestId || getTestId(ComponentDefaultTestId.LINEAR_PROGRESS_BAR, id)}
      >
        <div className={styles.container}>
          {renderBaseBars}
          {renderMultiBars}
        </div>
        {renderPercentage}
      </div>
    );
  }
);

export default withStaticProps(LinearProgressBar, {
  styles: ProgressBarStyleEnum,
  barStyles: ProgressBarStyleEnum,
  types: ProgressBarTypeEnum,
  barTypes: ProgressBarTypeEnum,
  sizes: SIZES
});
