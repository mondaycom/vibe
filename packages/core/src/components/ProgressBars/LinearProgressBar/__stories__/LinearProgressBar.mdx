import { Meta } from "@storybook/blocks";
import {
  SKELETON,
  SPINNER,
  STEPS
} from "../../../../storybook/components/related-components/component-description-map";
import {
  ComponentRuleNegative,
  ComponentRulePositive,
  TipMultiStepIndicator
} from "./LinearProgressBar.stories.helpers";
import * as LinearProgressBarStories from "./LinearProgressBar.stories";

<Meta of={LinearProgressBarStories} />

# LinearProgressBar

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Variants](#variants)
- [Do’s and don’ts](#dos-and-donts)
- [Use cases and examples](#use-cases-and-examples)
- [Related components](#related-components)
- [Feedback](#feedback)

## Overview

Progress bars show continuous progress through a process, such as a percentage value. They show how much progress is complete and how much remains.

<Canvas of={LinearProgressBarStories.Overview} />

### Import

```js
import { LinearProgressBar } from "@vibe/core";
```

## Props

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={["Give users an indication of how much of the task has been completed and how much is left."]}
/>

<TipMultiStepIndicator />

## Variants

### Regular

<Canvas of={LinearProgressBarStories.Regular} />

### With secondary value

<Canvas of={LinearProgressBarStories.WithSecondaryValue} />

### Multi progress bar

<Canvas of={LinearProgressBarStories.MultiProgressBar} />

## Do’s and Don’ts

<ComponentRules
  rules={[
    {
      positive: {
        component: <ComponentRulePositive />,
        description: "Use a progress bar only process has start and finish point."
      },
      negative: {
        component: <ComponentRuleNegative />,
        description: "Don’t use an infinite scalable indicator."
      }
    }
  ]}
/>

## Use cases and examples

### Progress bar as a counter

The user can see in a clear way the number of items used in the account.

<Canvas of={LinearProgressBarStories.ProgressBarAsACounter} />

### Progress bar as loading indicator

<Canvas of={LinearProgressBarStories.ProgressBarAsLoadingIndicator} />

## Related components

<RelatedComponents componentsNames={[SPINNER, STEPS, SKELETON]} />
