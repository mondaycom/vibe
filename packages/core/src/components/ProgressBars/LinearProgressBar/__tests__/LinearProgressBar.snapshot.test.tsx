import { describe, it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import LinearProgressBar from "../LinearProgressBar";

const multiValues = [
  { value: 25, color: "red" },
  { value: 75, color: "yellow" },
  { value: 100, color: "green" }
];

describe("LinearProgressBar renders correctly", () => {
  it("with empty props", () => {
    const tree = renderer.create(<LinearProgressBar />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with value", () => {
    const tree = renderer.create(<LinearProgressBar value={20} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with size", () => {
    const tree = renderer.create(<LinearProgressBar value={20} size="large" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with barStyle", () => {
    const tree = renderer.create(<LinearProgressBar value={20} barStyle="positive" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with min", () => {
    const tree = renderer.create(<LinearProgressBar value={20} min={10} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with indicateProgress", () => {
    const tree = renderer.create(<LinearProgressBar value={20} indicateProgress />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with valueSecondary", () => {
    const tree = renderer.create(<LinearProgressBar value={20} valueSecondary={40} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with multi and multiValues", () => {
    const tree = renderer.create(<LinearProgressBar value={20} multi multiValues={multiValues} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with full width progress bar", () => {
    const tree = renderer.create(<LinearProgressBar value={20} fullWidth />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
