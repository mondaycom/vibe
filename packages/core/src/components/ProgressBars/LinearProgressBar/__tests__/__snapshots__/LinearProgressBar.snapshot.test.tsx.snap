// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LinearProgressBar renders correctly > with barStyle 1`] = `
<div
  className="wrapper small"
  data-testsid="linear-progress-bar"
>
  <div
    className="container"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={20}
      className="bar typePrimaryPositive progressBar animate"
      data-testid="bar-primary"
      role="progressbar"
      style={
        {
          "width": "20%",
        }
      }
    />
  </div>
</div>
`;

exports[`LinearProgressBar renders correctly > with empty props 1`] = `
<div
  className="wrapper small"
  data-testsid="linear-progress-bar"
>
  <div
    className="container"
  />
</div>
`;

exports[`LinearProgressBar renders correctly > with full width progress bar 1`] = `
<div
  className="wrapper small fullWidth"
  data-testsid="linear-progress-bar"
>
  <div
    className="container"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={20}
      className="bar typePrimaryPrimary progressBar animate"
      data-testid="bar-primary"
      role="progressbar"
      style={
        {
          "width": "20%",
        }
      }
    />
  </div>
</div>
`;

exports[`LinearProgressBar renders correctly > with indicateProgress 1`] = `
<div
  className="wrapper small"
  data-testsid="linear-progress-bar"
>
  <div
    className="container"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={20}
      className="bar typePrimaryPrimary progressBar animate"
      data-testid="bar-primary"
      role="progressbar"
      style={
        {
          "width": "20%",
        }
      }
    />
  </div>
  <label
    className="label"
    htmlFor="linear-progress-bar"
  >
    20%
  </label>
</div>
`;

exports[`LinearProgressBar renders correctly > with min 1`] = `
<div
  className="wrapper small"
  data-testsid="linear-progress-bar"
>
  <div
    className="container"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={11.11111111111111}
      className="bar typePrimaryPrimary progressBar animate"
      data-testid="bar-primary"
      role="progressbar"
      style={
        {
          "width": "11.11111111111111%",
        }
      }
    />
  </div>
</div>
`;

exports[`LinearProgressBar renders correctly > with multi and multiValues 1`] = `
<div
  className="wrapper small"
  data-testsid="linear-progress-bar"
>
  <div
    className="container"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={100}
      className="bar typePrimaryNone progressBar completed animate"
      data-testid="bar_bar_green_0"
      id="bar_green_0"
      role="progressbar"
      style={
        {
          "backgroundColor": "green",
          "width": "100%",
        }
      }
    />
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={75}
      className="bar typePrimaryNone progressBar animate"
      data-testid="bar_bar_yellow_1"
      id="bar_yellow_1"
      role="progressbar"
      style={
        {
          "backgroundColor": "yellow",
          "width": "75%",
        }
      }
    />
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={25}
      className="bar typePrimaryNone progressBar animate"
      data-testid="bar_bar_red_2"
      id="bar_red_2"
      role="progressbar"
      style={
        {
          "backgroundColor": "red",
          "width": "25%",
        }
      }
    />
  </div>
</div>
`;

exports[`LinearProgressBar renders correctly > with size 1`] = `
<div
  className="wrapper large"
  data-testsid="linear-progress-bar"
>
  <div
    className="container"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={20}
      className="bar typePrimaryPrimary progressBar animate"
      data-testid="bar-primary"
      role="progressbar"
      style={
        {
          "width": "20%",
        }
      }
    />
  </div>
</div>
`;

exports[`LinearProgressBar renders correctly > with value 1`] = `
<div
  className="wrapper small"
  data-testsid="linear-progress-bar"
>
  <div
    className="container"
  >
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={20}
      className="bar typePrimaryPrimary progressBar animate"
      data-testid="bar-primary"
      role="progressbar"
      style={
        {
          "width": "20%",
        }
      }
    />
  </div>
</div>
`;

exports[`LinearProgressBar renders correctly > with valueSecondary 1`] = `
<div
  className="wrapper small"
  data-testsid="linear-progress-bar"
>
  <div
    className="container"
  >
    <div
      aria-label=""
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={40}
      className="bar typeSecondaryPrimary progressBar animate"
      data-testid="bar-secondary"
      role="progressbar"
      style={
        {
          "width": "40%",
        }
      }
    />
    <div
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={20}
      className="bar typePrimaryPrimary progressBar animate"
      data-testid="bar-primary"
      role="progressbar"
      style={
        {
          "width": "20%",
        }
      }
    />
  </div>
</div>
`;
